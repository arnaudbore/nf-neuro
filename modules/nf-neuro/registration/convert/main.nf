process REGISTRATION_CONVERT {
    tag "$meta.id"
    label 'process_single'

    container "freesurfer/freesurfer:7.4.1"
    containerOptions "--env FSLOUTPUTTYPE='NIFTI_GZ'"

    input:
    tuple val(meta), path(transformation), val(input_type), val(output_type), path(reference), path(affine_source), path(fs_license)

    output:
    tuple val(meta), path("*out_{affine,warp}.{nii,nii.gz,mgz,m3z,txt,lta,mat,dat}")    , emit: transformation
    path "versions.yml"                                                                 , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"

    def transform_types = [
        affine: [lta: "lta", fsl: "mat", mni: "xfm", reg: "dat", niftyreg: "txt", itk: "txt", vox: "txt"],
        warp: [m3z: "m3z", fsl: "nii.gz", lps: "nii.gz", itk: "nii.gz", ras: "nii.gz", vox: "mgz"]
    ]
    def spaces = ["ras2ras", "vox2vox", "register_dat"]

    // Validation transformation type and coercion with conversion type
    def in_extension = transformation.name.tokenize('.')[1..-1].join('.')
    def transform_type = transform_types.affine.find{ it.value == in_extension } ? "affine" : "warp"
    if ( transform_type == "warp" && !transform_types.warp.containsKey(output_type) ) {
        error "Invalid combination of transformation type and conversion type: ${transform_type} to ${output_type}."
    }

    def out_extension = transform_types[transform_type][output_type]
    def output_name = "${prefix}_out_${transform_type}.${out_extension}"
    def command = transform_type == "affine" ? "lta_convert" : "mri_warp_convert"

    if ( transform_type == "affine" ) {
        // Affine transformations are defined on the target space
        args += " --in$input_type $transformation --out$output_type $output_name --trg $reference"

        // Validate source geometry is available for conversion to .lta
        if ( output_type == "lta" ) {
            if ( !affine_source ) error "Source geometry must be provided for conversion to .lta."
            args += " --src $affine_source"
        }

        if ( task.ext.invert ) args += " --invert"
        if ( task.ext.conform ) args += " --trgconform"
        if ( task.ext.output_space ) {
            if ( !spaces.contains(task.ext.output_space) ) {
                error "Invalid output space: ${task.ext.output_space}. Valid options are: ${spaces.join(', ')}"
            }
            args += " --outspace " + task.ext.output_space
        }
    }
    else {
        // Deformable transformations are defined on the source space
        args += " --insrcgeom $reference --in$input_type $transformation --out$output_type $output_name"

        if ( task.ext.downsample ) args += " --downsample"
    }
    """
    export ITK_GLOBAL_DEFAULT_NUMBER_OF_THREADS=$task.cpus
    export OMP_NUM_THREADS=1
    export OPENBLAS_NUM_THREADS=1

    cp $fs_license \$FREESURFER_HOME/license.txt

    $command $args

    rm \$FREESURFER_HOME/license.txt

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        freesurfer: \$(mri_convert -version | grep "freesurfer" | sed -E 's/.* ([0-9.]+).*/\\1/')
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    def transform_types = [
        affine: [lta: "lta", fsl: "mat", mni: "xfm", reg: "dat", niftyreg: "txt", itk: "txt", vox: "txt"],
        warp: [m3z: "m3z", fsl: "nii.gz", lps: "nii.gz", itk: "nii.gz", ras: "nii.gz", vox: "mgz"]
    ]
    // Validation transformation type and coercion with conversion type
    def in_extension = transformation.name.tokenize('.')[1..-1].join('.')
    def transform_type = transform_types.affine.find{ it.value == in_extension } ? "affine" : "warp"
    if ( transform_type == "warp" && !transform_types.warp.containsKey(output_type) ) {
        error "Invalid combination of transformation type and conversion type: ${transform_type} to ${output_type}."
    }

    def out_extension = transform_types[transform_type][output_type]
    def output_name = "${prefix}_out_${transform_type}.${out_extension}"
    """
    set +e
    function handle_code () {
        local code=\$?
        ignore=( 1 )
        [[ " \${ignore[@]} " =~ " \$code " ]] || exit \$code
    }
    trap 'handle_code' ERR

    lta_convert --help
    mri_warp_convert --help

    touch $output_name

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        freesurfer: \$(mri_convert -version | grep "freesurfer" | sed -E 's/.* ([0-9.]+).*/\\1/')
    END_VERSIONS
    """
}
