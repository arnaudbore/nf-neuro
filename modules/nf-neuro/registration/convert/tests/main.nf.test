nextflow_process {

    name "Test Process REGISTRATION_CONVERT"
    script "../main.nf"
    process "REGISTRATION_CONVERT"
    config "./nextflow.config"

    tag "modules"
    tag "modules_nfneuro"
    tag "registration"
    tag "registration/convert"

    tag "subworkflows"
    tag "subworkflows/load_test_data"


    setup {
        run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
            script "../../../../../subworkflows/nf-neuro/load_test_data/main.nf"
            process {
                """
                input[0] = Channel.from( [ "freesurfer.zip" , "freesurfer_reslice.zip" , "freesurfer_transforms.zip"] )
                input[1] = "test.load-test-data"
                """
            }
        }
    }

    test("registration - convert - deformation - ras to itk (lps)") {
        when {
            process {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        freesurfer: it.simpleName == "freesurfer"
                        reslice: it.simpleName == "freesurfer_reslice"
                        transforms: it.simpleName == "freesurfer_transforms"
                    }
                ch_transform = ch_split_test_data.transforms.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/fs_deform.nii.gz")
                    ]
                }
                ch_reference = ch_split_test_data.reslice.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/t1_reslice.nii.gz")
                    ]
                }
                ch_license = ch_split_test_data.freesurfer.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/license.txt")
                    ]
                }
                input[0] = ch_transform
                    .join(ch_reference)
                    .join(ch_license)
                    .map{ meta, transform, reference, license -> [ meta, transform, "ras", "lps", reference, [], license ] }
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("registration - convert - affine - lta to fsl") {
        when {
            process {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        freesurfer: it.simpleName == "freesurfer"
                        reslice: it.simpleName == "freesurfer_reslice"
                        transforms: it.simpleName == "freesurfer_transforms"
                    }
                ch_transform = ch_split_test_data.transforms.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/fs_affine.lta")
                    ]
                }
                ch_reference = ch_split_test_data.reslice.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/fa_reslice.nii.gz")
                    ]
                }
                ch_license = ch_split_test_data.freesurfer.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/license.txt")
                    ]
                }
                input[0] = ch_transform
                    .join(ch_reference)
                    .join(ch_license)
                    .map{ meta, transform, reference, license -> [ meta, transform, "lta", "fsl", reference, [], license ] }
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("registration - convert - mixed - to itk") {
        when {
            process {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        freesurfer: it.simpleName == "freesurfer"
                        reslice: it.simpleName == "freesurfer_reslice"
                        transforms: it.simpleName == "freesurfer_transforms"
                    }
                ch_transforms = ch_split_test_data.transforms.map{
                    test_data_directory -> [
                        [[ id:'test1' ], [ id:'test2' ]],
                        [
                            file("\${test_data_directory}/fs_deform.nii.gz"),
                            file("\${test_data_directory}/fs_affine.lta")
                        ],
                        [ "ras", "lta" ],
                        [ "itk" ]
                    ]
                }
                ch_reference = ch_split_test_data.reslice.map{
                    test_data_directory -> [
                        [[ id:'test1' ], [ id:'test2' ]],
                        [
                            file("\${test_data_directory}/t1_reslice.nii.gz"),
                            file("\${test_data_directory}/fa_reslice.nii.gz")
                        ]
                    ]
                }
                ch_license = ch_split_test_data.freesurfer.map{
                    test_data_directory -> file("\${test_data_directory}/license.txt")
                }
                input[0] = ch_transforms
                    .join(ch_reference)
                    .transpose()
                    .combine(ch_license)
                    .map{ meta, transform, intype, outtype, reference, license ->
                        [ meta, transform, intype, outtype, reference, [], license ]
                    }
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("registration - convert - affine - to lta (needs source)") {
        when {
            process {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        freesurfer: it.simpleName == "freesurfer"
                        reslice: it.simpleName == "freesurfer_reslice"
                        transforms: it.simpleName == "freesurfer_transforms"
                    }
                ch_transform = ch_split_test_data.transforms.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/fsl_affine.mat")
                    ]
                }
                ch_reference = ch_split_test_data.reslice.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/fa_reslice.nii.gz")
                    ]
                }
                ch_source = ch_split_test_data.reslice.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/t1_reslice.nii.gz")
                    ]
                }
                ch_license = ch_split_test_data.freesurfer.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/license.txt")
                    ]
                }
                input[0] = ch_transform
                    .map{ meta, transform -> [meta, transform, "fsl", "lta"] }
                    .join(ch_reference)
                    .join(ch_source)
                    .join(ch_license)
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    file(process.out.transformation.get(0).get(1)).name,
                    process.out.versions
                ).match() }
            )
        }
    }

    test("registration - convert - stub-run") {
        tag "stub"
        options "-stub-run"

        when {
            process {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        freesurfer: it.simpleName == "freesurfer"
                        reslice: it.simpleName == "freesurfer_reslice"
                        transforms: it.simpleName == "freesurfer_transforms"
                    }
                ch_transform = ch_split_test_data.transforms.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/fs_deform.nii.gz"),
                    ]
                }
                ch_reference = ch_split_test_data.reslice.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/t1_reslice.nii.gz"),
                    ]
                }
                ch_license = ch_split_test_data.freesurfer.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/license.txt")
                    ]
                }
                input[0] = ch_transform
                    .join(ch_reference)
                    .join(ch_license)
                    .map{ meta, transform, reference, license -> [ meta, transform, "ras", "itk", reference, [], license ] }
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out.versions).match() }
            )
        }
    }
}
