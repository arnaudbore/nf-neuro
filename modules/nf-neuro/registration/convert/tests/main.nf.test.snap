{
    "registration - convert - stub-run": {
        "content": [
            [
                "versions.yml:md5,8fa42279d7793bb057cd5680bcb241e7"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-02T20:44:37.839613055"
    },
    "registration - convert - affine - to lta (needs source)": {
        "content": [
            "test_out_affine.lta",
            [
                "versions.yml:md5,8fa42279d7793bb057cd5680bcb241e7"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.8"
        },
        "timestamp": "2025-10-16T15:03:48.505663297"
    },
    "registration - convert - affine - lta to fsl": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_out_affine.mat:md5,c88fcae39c94d8af95ca00592c2f2634"
                    ]
                ],
                "1": [
                    "versions.yml:md5,8fa42279d7793bb057cd5680bcb241e7"
                ],
                "transformation": [
                    [
                        {
                            "id": "test"
                        },
                        "test_out_affine.mat:md5,c88fcae39c94d8af95ca00592c2f2634"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,8fa42279d7793bb057cd5680bcb241e7"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.10.0"
        },
        "timestamp": "2025-10-29T11:13:34.703328482"
    },
    "registration - convert - mixed - to itk": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test1"
                        },
                        "test1_out_warp.nii.gz:md5,69052e6226da946bad1f9466285cbb89"
                    ]
                ],
                "1": [
                    "versions.yml:md5,8fa42279d7793bb057cd5680bcb241e7"
                ],
                "transformation": [
                    [
                        {
                            "id": "test1"
                        },
                        "test1_out_warp.nii.gz:md5,69052e6226da946bad1f9466285cbb89"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,8fa42279d7793bb057cd5680bcb241e7"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.10.0"
        },
        "timestamp": "2025-10-29T11:13:43.762952295"
    },
    "registration - convert - deformation - ras to itk (lps)": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_out_warp.nii.gz:md5,69052e6226da946bad1f9466285cbb89"
                    ]
                ],
                "1": [
                    "versions.yml:md5,8fa42279d7793bb057cd5680bcb241e7"
                ],
                "transformation": [
                    [
                        {
                            "id": "test"
                        },
                        "test_out_warp.nii.gz:md5,69052e6226da946bad1f9466285cbb89"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,8fa42279d7793bb057cd5680bcb241e7"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.10.0"
        },
        "timestamp": "2025-10-29T11:13:25.855140969"
    }
}